source/Halibut.Tests/Support/SerilogLoggerBuilder.cs

```diff
@@ -14,58 +14,69 @@ namespace Halibut.Tests.Support
     {
         static readonly ILogger Logger;
+        static readonly ConcurrentDictionary<string, TraceLogFileLogger> TraceLoggers = new();
+
+        TraceLogFileLogger? traceFileLogger;
 
         static SerilogLoggerBuilder()
         {
-            var outputTemplate = 
-                "{TestName} "
+            const string teamCityOutputTemplate =
+                "{TestHash} "
                 + "{Timestamp:HH:mm:ss.fff zzz} "
                 + "{ShortContext} "
                 + "{Message}{NewLine}{Exception}";
 
+            const string localOutputTemplate =
+                "{Timestamp:HH:mm:ss.fff zzz} "
+                + "{ShortContext} "
+                + "{Message}{NewLine}{Exception}";
+
+            var nUnitOutputTemplate = TeamCityDetection.IsRunningInTeamCity()
+                ? teamCityOutputTemplate
+                : localOutputTemplate;
+
             Logger = new LoggerConfiguration()
-                .MinimumLevel.Debug()
-                .WriteTo.Sink(new NonProgressNUnitSink(new MessageTemplateTextFormatter(outputTemplate)))
+                .MinimumLevel.Verbose()
+                .WriteTo.Sink(new NonProgressNUnitSink(new MessageTemplateTextFormatter(nUnitOutputTemplate)), LogEventLevel.Debug)
+                .WriteTo.Sink(new TraceLogsForFailedTestsSink(new MessageTemplateTextFormatter(localOutputTemplate)))
                 .CreateLogger();
         }
 
+        public SerilogLoggerBuilder SetTraceLogFileLogger(TraceLogFileLogger logger)
+        {
+            this.traceFileLogger = logger;
+            return this;
+        }
+
         public ILogger Build()
         {
             // In teamcity we need to know what test the log is for, since we can find hung builds and only have a single file containing all log messages.
-            var testName = "";
-            if (TeamCityDetection.IsRunningInTeamCity())
-            {
-                testName = CurrentTestHash();
-            }
+            var testName = TestContext.CurrentContext.Test.FullName;
+            var testHash = CurrentTestHash();
+            var logger = Logger.ForContext("TestHash", testHash);
 
-            var logger = Logger.ForContext("TestName", testName);
-            
-            if (TeamCityDetection.IsRunningInTeamCity())
+            if (traceFileLogger != null)
             {
-                if (!HasLoggedTestHash.Contains(TestContext.CurrentContext.Test.Name))
-                {
-                    HasLoggedTestHash.Add(TestContext.CurrentContext.Test.Name);
-                    logger.Information($"Test: {TestContext.CurrentContext.Test.Name} has hash {CurrentTestHash()}");
-                }
+                TraceLoggers.AddOrUpdate(testName, traceFileLogger, (_, _) => throw new Exception("This should never be updated. If it is, it means that a test is being run multiple times in a single test run"));
+                logger.Information($"Test: {TestContext.CurrentContext.Test.Name} has hash {testHash}");
```

Comment:
Before, we would log this always (on team city). Can we get into a situation where `traceFileLogger` is null and we want to log this?

Reply:
Hmmmm, yes actually - we won't log this line anymore if:
* Test fixture doesn't derive from `BaseTest`
* `SerilogLoggerBuilder` is called manually to log some lines, but without initializing a trace logger.

This means it might be hard to find non-`BastTest` tests in the TC build log, as there is no hash printed in the test's logs.

The motivation for moving this line into this `if` block was that we only want this line printed once per test, but it would get printed multiples times if we print it everytime `Build` is called.

I'll look into some other options to deal with this scenario.

Reply:
I'm going to re-add the `HasLoggedTestHash` concurrent list, as this is the reason it existed - I was trying to consolidate it with the dictionary of trace loggers, but it serves it's own purpose so we need both. 🤷🏻